### ISSUE TITLE
Dlib train human face landmarks stop with no output

### ISSUE THREAD

**Issue by jiangwenyou** [davisking/dlib#1548]
Question Description:

I trained 83 key points of face. After several hours of training, the training process seemed to stop calculating, and then the console showed:

Fitting trees...
Time remaining: 1.96 hours.

At this time, the CPU usage of the training process is 0%, and the memory usage has not changed. It seems that it is not working, but it is still running in the task manager.

My running environment:

Windows10 x64
Visual studio c++ complier, build release version
CPU Intel xcon 2.4gx2 12core 24threads
RAM 24GB
Dlib version 19.16

All training parameters are the default, except for the following:

Trainer.set_nu (0.05);
Trainer.set_tree_depth (2);
Trainer.set_oversampling_amount (10);
Trainer.set_num_threads (24);

Train images count:~65000

This happens occasionally. After many attempts, I found that if the training set is very small, the training time is very short, the training will end soon, the training will be successful, and get sp. dat. But when the training set is large and the training time is very long, it will unpredictability stop at a certain time point.

I don't know what the reason is, there is no hint on the console. I'm sure my computer's CPU and RAM are enough. When I try it on two computers, there will be similar situations.

Thanks a lot!

**Comment by davisking** [comment]
That's a lot of images, what makes you sure you aren't running out of RAM
and the process isn't just swapping to disk over and over?

**Comment by jiangwenyou** [comment]
> 
> That's a lot of images, what makes you sure you aren't running out of RAM and the process isn't just swapping to disk over and over?

Just now, the training stopped again, and the task manager showed that there was still 8.8G memory left. Why? Is there enough memory?

![1](https://user-images.githubusercontent.com/6021263/48328011-ed843d80-e67c-11e8-9b23-67f2d174872c.PNG)
![2](https://user-images.githubusercontent.com/6021263/48328014-efe69780-e67c-11e8-9aa5-088cdab18f4c.PNG)

**Comment by davisking** [comment]
That's weird.  There isn't much going on in the trainer, it's just some for loops.  It's not like it's blocking on anything.  Are you sure you haven't modified the code or done something to cause it to pause?

**Comment by jiangwenyou** [comment]
> 
> 
> That's weird. There isn't much going on in the trainer, it's just some for loops. It's not like it's blocking on anything. Are you sure you haven't modified the code or done something to cause it to pause?


[MyTrain.hpp.txt](https://github.com/davisking/dlib/files/2572213/MyTrain.hpp.txt)

Thank you for your reply. This file is my training code. I don't know what's wrong. After a long training period, I stopped suddenly. It's really frustrating. Where is the problem? The data format of the training is certainly not a problem, because I have successfully trained on small data sets. I have tried with two computers, and this happens.



After many attempts, it has been found that training always stops at the end of training, for example, 15 hours of training will stop at about 2 hours from the end of training; if training lasts for 6 hours, it will stop at about 1 hour from the end of training.


When the training process is started, the console does not have any output, but the training process is occupying more and more memory until at some point, the memory occupancy begins to decline, and the console will display:


Fitting trees...


One more minute, the console will show:


Time remaining: XXX hours.


After that, the CPU will fluctuate between 10% and 70% in the training process, the memory occupancy will also change, and the remaining training time will continue to decrease.


Towards the end, it should have been exciting, but the search process stopped, the CPU occupied 0%, and the training time did not change. It's very frustrating!
